### DISCUSSION TITLE
Mocking Restease

### DISCUSSION THREAD

**Discussion by aadvark-pancakes** [canton7/RestEase#199] [Q&A] [answered]
Hi,
I'm relatively new to restease but I'm very happy with it. However I'm having trouble writing unit tests, more specifically **mocking** it out. 
I'm writing a small library which handles communication with an external media manager. Relatively simple library: get assets, update metadata, create new files, etc.

The specific trouble I'm having is mocking the calls to an endpoint. **How can I mock the restease client** so I can unit test it, instead of having to pass a real endpoint?

I realise I might be asking the wrong question, but any suggestions / nudges in the right direction would be appreciated. I'm quite new to this so I'm still trying to wrap my head around it.

I'm using moq and nunit. The example below is just a simplified version of what I'm trying.

Eg:

```
    public interface IBrandFolderApi
    {
        [Post("services/apilogin")]
        Task<LoginResponse> Login([Query] string username, [Query] string password);
    }


    public class BrandfolderClient : IBrandfolderClient
    {
        private IBrandFolderApi _brandFolderApi { get; set; } 

        public BrandfolderClient(string url)
        {
            _brandFolderApi = RestClient.For<IBrandFolderApi >(url);
        }

        public async Task<string> Login(string username, string password)
        {
            LoginResponse loginResponse = await _brandFolderApi .Login(username, password);
            if (loginResponse.LoginSuccess)
            {
                ....
            }
            ....            
            return loginResponse.LoginSuccess.ToString();
        }
}


    public class BrandFolderTests
    {
        BrandfolderClient  _brandfolderClient 
        Mock<IBrandFolderApi> _mockBrandFolderApii;


        [SetUp]
        public void Setup()
        {
             //The test will fail here, as I'm passing a real URL and it will try and contact it.
            //If I try and send any string, I receive an Invalid URL Format exception.
             string url = "https://brandfolder.companyname.io";
            _brandfolderClient = new BrandfolderClient  (url);
            _mockBrandFolderApii= new Mock<IBrandFolderApi>();
        }
        ....
    }

```

**Comment by canton7** [comment, marked as answer]
Your BrandfolderClient constructor does:

```cs
_brandFolderApi = RestClient.For<IBrandFolderApi >(url);
```

So your BrandfolderClient will always use a "real" IBrandFolderApi, and the mock you create in your unit test is unrelated to anything.

This is general programming help unrelated to RestEase, but if you want to test BrandfolderClient, you'll need a way to get it to use a mock IBrandFolderApi, rather than always calling `RestClient.For<IBrandFolderApi >(url)`. That's normally done by having a constructor which accepts an instance of `IBrandFolderApi`: this might be the only constructor, or you might have one constructor which takes a string, and one which takes a `IBrandFolderApi`.

**Reply by aadvark-pancakes** [reply]
Yes you're right, seems simple now. I've done exactly that and got it. I'm separating the concrete instance of RestClient.For<IBrandFolderApi >(url) from my library. This will be done in the code that uses the library. Thank you!
